### PR TITLE
Rework entry point selector.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/EntryPoint.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    entryPointDescriptor.bind(entryPoint)
+    entryPoint.annotations += buildSimpleAnnotation(context.irBuiltIns,
+            actualMain.startOffset, actualMain.startOffset,
+            context.ir.symbols.exportForCppRuntime.owner, "Konan_start")
+
+    val builder = context.createIrBuilder(entryPoint.symbol)
+    entryPoint.body = builder.irBlockBody(entryPoint) {
+        +IrTryImpl(
+                startOffset = actualMain.startOffset,
+                endOffset   = actualMain.startOffset,
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/EntryPoint.kt]: Is this intended?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ToplevelPhases.kt [hunk 2]
```diff
@@ -316,6 +318,9 @@ internal val dependenciesLowerPhase = SameTypeNamedPhaseWrapper(
                         }
                 input.files += files
 
+                if (context.config.produce == CompilerOutputKind.PROGRAM)
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ToplevelPhases.kt]: Adding entry point seems unrelated to lowering dependencies. Maybe extract explicit phase for this?

File: runtime/src/main/kotlin/kotlin/native/internal/RuntimeUtils.kt [hunk 3]
```diff
@@ -180,4 +180,13 @@ internal fun <T> listOfInternal(vararg elements: T): List<T> {
     for (i in 0 until elements.size)
         result.add(elements[i])
     return result
-}
\ No newline at end of file
+}
+
+
+@PublishedApi
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/kotlin/native/internal/RuntimeUtils.kt]: Why not have a single function here?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/EntryPoint.kt (hunk 1 above)
Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/EntryPoint.kt]: Yes, here and below we just set entry point as if it was in very beginning of the main.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ToplevelPhases.kt (hunk 2 above)
Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ToplevelPhases.kt]: Well, not sure if it shall a phase on its own. WDYT?

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ToplevelPhases.kt]: What do you mean?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ToplevelPhases.kt [hunk 4]
```diff
@@ -320,6 +322,20 @@ internal val dependenciesLowerPhase = SameTypeNamedPhaseWrapper(
...
 
+internal val entryPointPhase = SameTypeNamedPhaseWrapper(
+        name = "addEntryPoint",
+        description = "Add entry point for program",
+        prerequisite = emptySet(),
+        lower = object : CompilerPhase<Context, IrModuleFragment, IrModuleFragment> {
+            override fun invoke(phaseConfig: PhaseConfig, phaserState: PhaserState<IrModuleFragment>,
+                                context: Context, input: IrModuleFragment): IrModuleFragment {
+                assert(context.config.produce == CompilerOutputKind.PROGRAM)
+                context.ir.symbols.entryPoint!!.owner.file.addChild(makeEntryPoint(context))
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ToplevelPhases.kt]: Btw, this will not work for entry point in a "cached" dependency. I guess we'll need to generate a synthetic file for this.

Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ToplevelPhases.kt]: Yeah, likely so.
